File: src/csoclassify/extract/detect.py
#------------------------------------------------------------------
# 내용 기반 포맷 감지 (설계서 CSO_HybridParse §라우팅)
#=> 확장자가 아니라 "파일 내용"으로 실제 포맷을 판별한다. 확장자가 틀리거나
#   위장돼 있어도(.txt 인데 실제 doc 등) 올바른 파서로 보내기 위함이다.
#    1) 앞 8바이트 매직으로 큰 갈래를 나눈다: %PDF / OLE(D0CF11E0) / ZIP(PK)
#    2) OLE·ZIP 계열은 매직이 서로 같으므로(예: doc·xls·ppt·hwp 모두 OLE) 컨테이너
#       "내부 스트림/엔트리"를 열어 세부 포맷을 구분한다.
#    3) 매직이 없으면 널바이트 유무로 텍스트/바이너리를 가른다.
#   반환 타입은 라우팅 키로 쓰인다(pdf/hwp/hwpx/doc/docx/xls/xlsx/ppt/pptx/text/…).
#------------------------------------------------------------------

import zipfile

# olefile 은 hwp5 용으로 이미 번들. 없으면 OLE 세부구분만 포기(→ 'ole').
try:
    import olefile
except ImportError:
    olefile = None

# 복합문서(OLE/CFB) 시그니처 — doc/xls/ppt/hwp 가 모두 이 매직으로 시작한다.
_OLE_MAGIC = b"\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1"
# ZIP 시그니처 — docx/xlsx/pptx/hwpx 가 모두 이 매직으로 시작한다.
_ZIP_MAGIC = b"PK\x03\x04"


#------------------------------------------------------------------
# OLE 내부 스트림으로 세부 포맷 구분
#=> doc/xls/ppt/hwp 는 매직이 같으므로 컨테이너 안의 대표 스트림 이름으로 가른다.
#    · FileHeader 스트림에 "HWP Document File" 시그니처 → hwp
#    · WordDocument → doc / Workbook|Book → xls / PowerPoint Document → ppt
#
# -in: path = OLE 파일 경로
#
# -out: str = "hwp"|"doc"|"xls"|"ppt"|"ole"(알 수 없는 OLE)
# -out: error = 없음(열기 실패 시 "unknown")
#------------------------------------------------------------------
def _detect_ole(path):
    try:
        ole = olefile.OleFileIO(path)
    except Exception:   # noqa: BLE001  (손상 OLE 등)
        return "unknown"
    try:
        top = {e[0] for e in ole.listdir()}
        # HWP5 는 FileHeader 스트림의 32바이트 시그니처로 확정한다.
        if "FileHeader" in top:
            try:
                if ole.openstream("FileHeader").read(32).startswith(b"HWP Document File"):
                    return "hwp"
            except Exception:   # noqa: BLE001
                pass
        if "WordDocument" in top:
            return "doc"
        if "Workbook" in top or "Book" in top:   # BIFF8/BIFF5
            return "xls"
        if "PowerPoint Document" in top:
            return "ppt"
        # 그 외 OLE(예: .msg) 는 전용 파서가 없다 → snf 폴백 대상.
        return "ole"
    finally:
        ole.close()


#------------------------------------------------------------------
# ZIP 내부 엔트리로 세부 포맷 구분
#=> docx/xlsx/pptx/hwpx 는 매직이 같으므로 컨테이너 안의 대표 파일로 가른다.
#
# -in: path = ZIP 파일 경로
#
# -out: str = "docx"|"xlsx"|"pptx"|"hwpx"|"zip"(일반 zip)
# -out: error = 없음(열기 실패 시 "unknown")
#------------------------------------------------------------------
def _detect_zip(path):
    try:
        with zipfile.ZipFile(path) as z:
            names = set(z.namelist())
            if "word/document.xml" in names:
                return "docx"
            if "xl/workbook.xml" in names:
                return "xlsx"
            if "ppt/presentation.xml" in names:
                return "pptx"
            # HWPX: Contents/section0.xml… 또는 mimetype 에 hwp 표기
            if any(n.startswith("Contents/section") for n in names):
                return "hwpx"
            if "mimetype" in names:
                try:
                    if b"hwp" in z.read("mimetype"):
                        return "hwpx"
                except Exception:   # noqa: BLE001
                    pass
        # 일반 zip 은 전용 파서 없음 → snf 폴백 대상.
        return "zip"
    except Exception:   # noqa: BLE001  (손상 zip 등)
        return "unknown"


#------------------------------------------------------------------
# 파일 → 포맷 타입 (핵심)
#=> 파일 내용으로 실제 포맷을 판별해 라우팅 키를 돌려준다(확장자 무시).
#    1) 앞 8바이트 읽기 → %PDF / OLE / ZIP 매직 판별
#    2) OLE·ZIP 은 내부를 열어 세부 구분(_detect_ole/_detect_zip)
#    3) 매직 없으면 널바이트 유무로 text/binary 판정
#
# -in: path = 대상 파일 경로
#
# -out: str = pdf|hwp|hwpx|doc|docx|xls|xlsx|ppt|pptx|text|ole|zip|binary|unknown
# -out: error = 없음(읽기 실패 등은 "unknown")
#------------------------------------------------------------------
def detect_format(path):
    try:
        with open(path, "rb") as f:
            # 앞부분을 넉넉히 읽는다: 매직(8B)뿐 아니라 HTML 마커(<html 등)도 보려면 필요.
            head = f.read(2048)
    except OSError:
        return "unknown"
    if not head:
        return "unknown"
    # PDF: 앞이 "%PDF-".
    if head[:5] == b"%PDF-":
        return "pdf"
    # 이미지: 전용 텍스트 파서가 없다(OCR 별도) → 'image'로 두고 snf 폴백/미분류.
    #   (PNG 는 널바이트가 없어 아래 text 판정에 잘못 걸리므로 여기서 먼저 잡는다.)
    if (head[:8] == b"\x89PNG\r\n\x1a\n" or head[:3] == b"\xFF\xD8\xFF"
            or head[:6] in (b"GIF87a", b"GIF89a") or head[:2] == b"BM"):
        return "image"
    # OLE 복합문서: 내부 스트림으로 세부 구분(olefile 없으면 'ole' 로 두고 snf 폴백).
    #   head 는 2048바이트라 8바이트 매직과 통째로 비교하면 절대 참이 되지 않는다.
    #   반드시 앞 8바이트만 잘라 비교한다(바로 아래 ZIP 판정과 같은 방식).
    if head[:8] == _OLE_MAGIC:
        return _detect_ole(path) if olefile is not None else "ole"
    # ZIP 컨테이너: 내부 엔트리로 세부 구분.
    if head[:4] == _ZIP_MAGIC:
        return _detect_zip(path)
    # 매직 없음: 널바이트가 없으면 텍스트 계열로 본다(txt/md/csv/json/ipynb/py/html 등).
    if b"\x00" not in head:
        # HTML 은 태그를 벗겨야 하므로 별도 타입으로 잡는다(앞부분에 html 마커가 있으면).
        low = head.lower()
        if (b"<!doctype html" in low or b"<html" in low
                or b"<head" in low or b"<body" in low):
            return "html"
        return "text"
    return "binary"

File: Rust/src/detect.rs
//! 내용 기반 포맷 감지 (Python detect.py 포팅)
//! 확장자가 아니라 매직바이트/컨테이너 내부 구조로 판별한다.

use std::fs::File;
use std::io::Read;
use std::path::Path;

/// 감지된 포맷 종류. Python detect_format 반환값과 동일 문자열 매핑.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Fmt {
    Pdf,
    Hwp,
    Hwpx,
    Doc,
    Docx,
    Xls,
    Xlsx,
    Ppt,
    Pptx,
    Html,
    Text,
    Image,
    Ole,
    Zip,
    Binary,
    Unknown,
}

impl Fmt {
    pub fn as_str(&self) -> &'static str {
        match self {
            Fmt::Pdf => "pdf", Fmt::Hwp => "hwp", Fmt::Hwpx => "hwpx",
            Fmt::Doc => "doc", Fmt::Docx => "docx", Fmt::Xls => "xls",
            Fmt::Xlsx => "xlsx", Fmt::Ppt => "ppt", Fmt::Pptx => "pptx",
            Fmt::Html => "html", Fmt::Text => "text", Fmt::Image => "image",
            Fmt::Ole => "ole", Fmt::Zip => "zip", Fmt::Binary => "binary",
            Fmt::Unknown => "unknown",
        }
    }
}

const OLE_MAGIC: &[u8] = &[0xD0, 0xCF, 0x11, 0xE0, 0xA1, 0xB1, 0x1A, 0xE1];
const ZIP_MAGIC: &[u8] = b"PK\x03\x04";

/// 파일 앞부분과 컨테이너 내부를 보고 포맷을 판별한다.
pub fn detect_format(path: &Path) -> Fmt {
    let mut head = [0u8; 2048];
    let n = match File::open(path).and_then(|mut f| f.read(&mut head)) {
        Ok(n) => n,
        Err(_) => return Fmt::Unknown,
    };
    let head = &head[..n];
    if head.is_empty() {
        return Fmt::Text;
    }

    // PDF
    if head.starts_with(b"%PDF-") {
        return Fmt::Pdf;
    }
    // 이미지 매직
    if head.starts_with(&[0x89, b'P', b'N', b'G', b'\r', b'\n', 0x1a, b'\n'])
        || head.starts_with(&[0xFF, 0xD8, 0xFF])
        || head.starts_with(b"GIF87a") || head.starts_with(b"GIF89a")
        || head.starts_with(b"BM")
    {
        return Fmt::Image;
    }
    // OLE(구형 doc/xls/ppt/hwp 공용) → 내부 스트림 검사
    if head.starts_with(OLE_MAGIC) {
        let quick = detect_ole(head);
        // 앞 2KB 안에 스트림명이 없으면(디렉터리가 파일 뒤쪽에 있는 흔한 경우)
        // 실제 CFB 디렉터리를 읽어 정확히 가린다. 이 폴백이 없으면 구형 .doc 이
        // 'ole' 로만 잡혀 추출기가 없는 포맷 취급을 받는다.
        if quick == Fmt::Ole {
            return detect_ole_exact(path);
        }
        return quick;
    }
    // ZIP(docx/xlsx/pptx/hwpx/일반 zip 공용) → 내부 이름 검사
    if head.starts_with(ZIP_MAGIC) {
        return detect_zip(path);
    }
    // HTML 마커
    let lower: Vec<u8> = head.iter().take(512).map(|b| b.to_ascii_lowercase()).collect();
    let ls = String::from_utf8_lossy(&lower);
    if ls.contains("<!doctype html") || ls.contains("<html") || ls.contains("<head") || ls.contains("<body") {
        return Fmt::Html;
    }
    // 텍스트 vs 바이너리: NUL 바이트 유무
    if head.contains(&0u8) {
        Fmt::Binary
    } else {
        Fmt::Text
    }
}

/// OLE 내부 스트림 이름을 UTF-16LE 흔적으로 대략 판별(경량 휴리스틱).
/// 완전한 CFB 파서 대신, 스트림명이 앞부분에 UTF-16LE 로 박혀 있는 점을 이용한다.
fn detect_ole(head: &[u8]) -> Fmt {
    // UTF-16LE 로 인코딩된 대표 스트림명을 바이트열로 찾는다.
    let has = |needle: &str| -> bool {
        let utf16: Vec<u8> = needle.encode_utf16().flat_map(|u| u.to_le_bytes()).collect();
        contains_subslice(head, &utf16)
    };
    if has("WordDocument") { return Fmt::Doc; }
    if has("Workbook") || has("Book") { return Fmt::Xls; }
    if has("PowerPoint Document") { return Fmt::Ppt; }
    if has("FileHeader") || has("HwpSummaryInformation") { return Fmt::Hwp; }
    Fmt::Ole
}

/// CFB 디렉터리를 실제로 읽어 OLE 세부 포맷을 가린다(위 휴리스틱이 실패했을 때만).
/// 파일 전체를 읽으므로 값이 비싸다 — 그래서 폴백으로만 쓴다.
fn detect_ole_exact(path: &Path) -> Fmt {
    let data = match std::fs::read(path) { Ok(d) => d, Err(_) => return Fmt::Ole };
    let ole = match crate::ole::Ole::open(data) { Some(o) => o, None => return Fmt::Ole };
    let names = ole.names();
    let has = |n: &str| names.iter().any(|x| *x == n);
    if has("WordDocument") { return Fmt::Doc; }
    if has("PowerPoint Document") { return Fmt::Ppt; }
    if has("Workbook") || has("Book") { return Fmt::Xls; }
    if has("FileHeader") || has("HwpSummaryInformation") { return Fmt::Hwp; }
    Fmt::Ole
}

/// ZIP 내부 항목 이름으로 OOXML/HWPX/일반 zip 을 구분.
fn detect_zip(path: &Path) -> Fmt {
    let file = match File::open(path) {
        Ok(f) => f,
        Err(_) => return Fmt::Zip,
    };
    let mut zip = match zip::ZipArchive::new(file) {
        Ok(z) => z,
        Err(_) => return Fmt::Zip,
    };
    let mut names: Vec<String> = Vec::with_capacity(zip.len());
    for i in 0..zip.len() {
        if let Ok(f) = zip.by_index(i) {
            names.push(f.name().to_string());
        }
    }
    let any = |pred: &dyn Fn(&str) -> bool| names.iter().any(|n| pred(n));
    if any(&|n| n == "word/document.xml") { return Fmt::Docx; }
    if any(&|n| n == "xl/workbook.xml") { return Fmt::Xlsx; }
    if any(&|n| n == "ppt/presentation.xml") { return Fmt::Pptx; }
    if any(&|n| n.starts_with("Contents/section")) || any(&|n| n == "Contents/content.hpf") {
        return Fmt::Hwpx;
    }
    Fmt::Zip
}

fn contains_subslice(hay: &[u8], needle: &[u8]) -> bool {
    if needle.is_empty() || needle.len() > hay.len() {
        return false;
    }
    hay.windows(needle.len()).any(|w| w == needle)
}
